Register projections and action plugin dynamic host/group/var API

nitzmahone opened the pull request:
## SUMMARY

### Register Projections

The functionality of the `register` keyword has been expanded to allow registration of multiple variables and inline manipulation of task output with Jinja expressions ("projections").

In addition to the existing `register: varname` form, the `register` keyword can accept a dictionary of `varname: expression` pairs (see examples below).

A new `_task` implicit variable provides Jinja expressions access to the current task's result via its `result` property. Expressions under looped tasks can also now access all completed loop results and preview the final loop task result via the `_task.loop_result` property. The new `_task` implicit variable is available to expressions in `register` and task conditionals (`changed_when`, `failed_when`, etc.). Registered variable values are updated and available to task conditional expressions as before, but task conditionals can now bypass `register` entirely by accessing `_task` directly (for tasks that do not otherwise require persistence of their result).

This functionality can simplify and declutter playbooks by allowing common patterns like "Run a task, 
register the output, pick out small parts of the output with `set_fact`" to be folded directly into the task itself. Further, because the new `register` form is always a Jinja expression, template delimiters (`{{ }}`)  are not required for dynamic behavior as they were with `set_fact`.

Unlike `set_fact`, variable registration expressions *can* refer to other variable expressions within the same task, and because they are evaluated lazily, can be defined in any order under `register` .


### Action APIs for creating dynamic hosts/groups/vars

A new  `register_host_variables` public API method on `ActionBase` allows user action plugins to implement variable-setting functionality that was previously restricted to built-in actions or required risky use of undocumented/internal action result values.

(TBD: expand description, usage examples)

The following built-in actions have been modified to use the new API:
* `set_fact`
* `include_vars`


## REGISTER PROJECTION EXAMPLES

#### Registering multiple variables from a task

**Before register-projections** 

```yml
- shell: echo hello
  failed_when: echo_result.stdout is contains "goodbye"
  register: echo_result

- set_fact:
    echo_duration: "{{ echo_result.delta }}"
    capitalized_result: "{{ echo_result.stdout | capitalize }}"
```

**After register-projections**

```yml
- shell: echo hello
  failed_when: _task.result.stdout is contains "goodbye"
  register:
    echo_result: _task.result  # The old register behavior
    echo_duration: _task.result.delta
    capitalized_result: _task.result.stdout | capitalize
```

#### Advanced templating and chaining variable manipulation

**Before register-projections** 

```yml
- shell: echo hello
  register: echo_output

- set_fact:
    start_hour: "{{ echo_output.start | regex_search('\\s(\\d{2}):\\d{2}:\\d{2}', '\\1') | first | int}}"

- set_fact:
    output_message: "{{ echo_output.stdout * start_hour }}"
```

**After register-projections** 

```yml
- shell: echo hello
  register:
    output_message: _task.result.stdout * start_hour  # Notice that start_hour is defined below, but can still be used! The order that they're declared does not matter. 
    start_hour: _task.result.start | regex_search('\\s(\\d{2}):\\d{2}:\\d{2}', '\\1') | first | int
```

#### Using `failed_when` without a register

**Before register-projections** 

```yml
- shell: echo hello
  failed_when: echo_result.stdout is contains "goodbye"
  register: echo_result
```

**After register-projections**

```yml
- shell: echo hello
  failed_when: _task.result.stdout is contains "goodbye"
```

#### Registering single loop items

**Before register-projections**

```yml
- shell: echo {{ item }}
  loop:
    - 1
    - 2
  register: echo_loop_result

- set_fact:
    second_echo_stdout: "{{ echo_loop_result.results[1].stdout }}"
```

**After register-projections**

```yml
- shell: echo {{ item }}
  loop:
    - 1
    - 2
  register:
    second_echo_stdout: _task.loop_results[1].stdout | default(0)  # using default here is necessary as when the loop is on the first item, it will still try and access `loop_results[1]` which doesn't exist yet
```

##### ISSUE TYPE

- Feature Pull Request

felixfontein commented:
I did a quick test of the vars setting API with `community.sops.load_vars`, and it worked fine.
```diff
diff --git a/plugins/action/load_vars.py b/plugins/action/load_vars.py
index 35087f4..5408624 100644
--- a/plugins/action/load_vars.py
+++ b/plugins/action/load_vars.py
@@ -103,8 +103,15 @@ class ActionModule(ActionModuleBase):
         if expressions == 'lazy-evaluation':
             value = self._make_safe(value)
 
-        module.exit_json(
-            ansible_included_var_files=files,
-            ansible_facts=value,
-            _ansible_no_log=True,
-        )
+        result = {
+            'ansible_included_var_files': files,
+            '_ansible_no_log': True,
+        }
+
+        if hasattr(self, 'register_host_variables'):
+            from ansible.plugins.action import VariableLayer
+            self.register_host_variables(variables=value, layer=VariableLayer.INCLUDE_VARS)
+        else:
+            result['ansible_facts'] = value
+
+        module.exit_json(**result)
```

nitzmahone committed:
Register projections and action plugin variable API

Co-authored-by: Matt Clay <matt@mystile.com>
Co-authored-by: Patrick Kingston <pkingsto@redhat.com>

nitzmahone committed:
misc post-rebase test fixes

mattclay committed:
Fix string field validation error context

mattclay committed:
Fix loop validation error handling

mattclay committed:
Fix pre-existing docstring typo

mattclay committed:
Remove outdated RPFIX comment

mattclay committed:
Add missing docstrings

mattclay committed:
Consolidate task pending changes and relocate to context

mattclay committed:
Remove RPFIX for pre-existing async_wrapper issue

mattclay committed:
Sanity test fixes

mattclay committed:
Drop `_` prefix from TaskContext's pending changes

mattclay committed:
Access Task from TE via TaskContext

Also move Task squashing outside TE.

mattclay committed:
Remove long-dead test code

mattclay committed:
Omit unused _ansible_delegated_vars from post_connection_vars

This has no effect on the same-named value in task results, which is used by callbacks.

mattclay committed:
Remove unncessary field reset

The field is not exportable, so there's no need to clear it to omit it from results.

mattclay committed:
Remove RPFIX for further Task consolidation

TE now uses Task from the context via a property.
Updating the worker makes less sense since it needs an instance attribute to pass Task from __init__ to run.

mattclay committed:
Remove unnecessary var deletion

The variable will be overwritten on the next loop iteration

mattclay committed:
Improve checks for loop_control conflicts

mattclay committed:
Remove obsolete var cleanup in TE

mattclay committed:
Switch FIXME to RPFIX

mattclay committed:
Add missing newline

mattclay committed:
Fix RP for skipped tasks and loop items

Added tests to cover these cases and a few others.
Also moved loop item label templating after result registration.

mattclay committed:
Relocate tests

mattclay committed:
Modular tests

mattclay committed:
More comprehensive RP tests

mattclay committed:
Simplify conditional

mattclay committed:
Remove obsolete comment

mattclay committed:
Remove redundant facts code

mattclay committed:
Temporarily disable changed tests for add_host/add_group

These will be fixed once the worker-side add_host/add_group changes are in place.

mattclay committed:
Update protomatter test

The test no longer relies on storing unsupported types during register for later assertion.
Instead, unmasked values are inspected within the task that generates them.

mattclay committed:
Make projections lazy

mattclay committed:
Remove allow_replace

It's a bigger trip-hazard than a help and is disabled most places it would have been useful.

mattclay committed:
Add test for missing async job id

mattclay committed:
Adjust pending changes behavior

They will be applied even if the action fails.
It is the responsibility of the action to not request changes unless they should be applied.

mattclay committed:
Sanity test fix

mattclay committed:
Adjust RPFIX comments

mattclay committed:
Fix updating ansible_facts in task vars during loop

mattclay committed:
Fix line length

mattclay committed:
Fix line length

mattclay committed:
Fix TE updates to task vars

mattclay committed:
Remove outdated RPFIX comment

mattclay committed:
Remove more outdated RPFIX comments

mattclay committed:
Expose warnings/deprecations during task exec

mattclay committed:
Add tests for warning context

s-hertel commented on test/integration/targets/ignore_errors/test_invalid_ignore_errors.yml:
Nothing is executing this file yet.

s-hertel commented on test/integration/targets/ignore_errors/test_invalid_ignore_errors.yml:
```diff
@@ -0,0 +1,29 @@
+#- hosts: localhost
+#  gather_facts: no
+#  tasks:
+#    - raw: exit {{ item }}
+#      ignore_errors: '{{ item }}'
+#      loop: [ 0,2,1 ]  # this doesn't really fail safe because any "true" means we ignore all errors - should an invalid ignore_errors in a loop be fully fatal to that host+task (or at least treated as ignore_errors: false for that host+task?)
```
I think the described behavior is a bug. The overall task should not ignore errors if any failed items are not ignored (in this case, the inability to interpret 2 as a boolean should be fatal). On both this branch and devel, the next task executes (not a new bug).

However, I don't think I understood what this branch was changing with `loop` + `ignore_errors`. Here's a variation of the same idea, except not a keyword processing error:

```yaml
- hosts: localhost
  gather_facts: no
  tasks:
    - raw: exit {{ item }}
      ignore_errors: '{{ item == 1 }}'
      loop: [ 0,2,1 ]

    - fail:
        msg: oops, should have exited?
```

Similarly to the test here, one of the two item failures are ignored, so I'd expect the overall task result to be failed.

<details><summary>On devel I get the expected behavior.</summary>

```console
PLAY [localhost] ********************************************************************************************************************************

TASK [raw] **************************************************************************************************************************************
changed: [localhost] => (item=0)
[ERROR]: Task failed: Action failed: non-zero return code
Origin: /home/shertel/ansible/test_ignore_errors.yml:4:7

2   gather_facts: no
3   tasks:
4     - raw: exit {{ item }}
        ^ column 7

failed: [localhost] (item=2) => {"ansible_loop_var": "item", "changed": true, "item": 2, "msg": "non-zero return code", "rc": 2, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
failed: [localhost] (item=1) => {"ansible_loop_var": "item", "changed": true, "item": 1, "msg": "non-zero return code", "rc": 1, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

PLAY RECAP **************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```
</details>

<details><summary>On this branch, the second task executes.</summary>

```console
PLAY [localhost] ********************************************************************************************************************************

TASK [raw] **************************************************************************************************************************************
changed: [localhost] => (item=0)
[ERROR]: Task failed: Action failed: non-zero return code
Origin: /home/shertel/ansible/test_ignore_errors.yml:4:7

2   gather_facts: no
3   tasks:
4     - raw: exit {{ item }}
        ^ column 7

failed: [localhost] (item=2) => {"ansible_loop_var": "item", "changed": true, "item": 2, "msg": "non-zero return code", "rc": 2, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
failed: [localhost] (item=1) => {"ansible_loop_var": "item", "changed": true, "item": 1, "msg": "non-zero return code", "rc": 1, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
...ignoring

TASK [fail] *************************************************************************************************************************************
[ERROR]: Task failed: Action failed: oops, should have exited?
Origin: /home/shertel/ansible/test_ignore_errors.yml:8:7

6       loop: [ 0,2,1 ]  # this doesn't really fail safe because any "true" means we ignore all errors - should an in...
7
8     - fail:
        ^ column 7

fatal: [localhost]: FAILED! => {"changed": false, "msg": "oops, should have exited?"}

PLAY RECAP **************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=1    skipped=0    rescued=0    ignored=1
```
</details>

mattclay committed:
RPFIX cleanup

mattclay committed:
Remove unreachable branch

mattclay committed:
Fix worker error handling

nitzmahone committed:
switch add_host/add_group to controller RPC

* eliminate duplicate controller-side result processing and callback dispatch for add_host/add_group
* implement LocalManager RPC server and dispatch utilities

Co-authored-by: Matt Clay <matt@mystile.com>

mattclay committed:
RPFIX cleanup

mattclay committed:
Restore add_host, changed_when, group_by tests

mattclay committed:
Remove unused Source.ACTION and DROP_AND_WARN

mattclay committed:
Use normal FA validation for break_when

mattclay committed:
Remove extraneous list conversion

Let normal FA processing handle changed_when and failed_when list conversion

mattclay committed:
Improve handling of break_when and changed_when

mattclay committed:
Remove unnecessary RPFIX

mattclay committed:
Drop priority on RPFIX

mattclay committed:
Replace RPFIX-1 with FUTURE for execute_meta

mattclay committed:
Add disabled test for RP-induced ignore_errors+loop bug

mattclay committed:
Enable test which was overlooked previously

mattclay committed:
Demote RPFIX-1 in test

mattclay committed:
Revert "Remove unused Source.ACTION and DROP_AND_WARN"

This reverts commit 80fdfbcbea398c17433fdc0402f33c339a4c7a23.

mattclay committed:
Remove RPFIX

mattclay committed:
Fix handling of invocation and _ansible_ prefixed result keys

Co-authored-by: Matt Davis <mrd@redhat.com>

mattclay committed:
Prepared for future removal of ansible_module_results hack

Co-authored-by: Matt Davis <mrd@redhat.com>

mattclay committed:
Cleaned up various RPFIX-1 comments

Co-authored-by: Matt Davis <mrd@redhat.com>

mattclay committed:
Fix yamllint issue

mattclay committed:
Fix callback_default test

mattclay committed:
Fix pass-through of callback-only fields

mattclay committed:
Remove outdated comment

mattclay committed:
Improve error context for changed_when, failed_when, break_when

mattclay committed:
Improve 'when' error display

mattclay committed:
Remove unnecessary 'else'

mattclay committed:
Improve until error messaging

mattclay committed:
Cleaned up remaining RPFIX-1 comments

The add_host/add_group API is now marked experimental.

Co-authored-by: Matt Davis <mrd@redhat.com>

mattclay commented:
/azp run

nitzmahone committed:
changelog cleanup
